RearrangeCarGroup8Movements.py

```diff
@@ -0,0 +1,206 @@
+        self.current_car_order[car_to_move_position] = 0
+
+        self.car_position_dict[car_to_move] = self.empty_position
+        self.empty_position = car_to_move_position
+
+    def _get_not_arranged_car_index(self):
+
+        '''Finds first car index not yet standing on the final destination
+        '''
+
+        car = random.choice(tuple(self.not_arranged_cars)) # O(1)
+        return self.car_position_dict[car]
+
+    def _rearrange_one_car(self):
+
+        '''Puts to the current empty space the car
+        that should be on this place according to the final arrangement
+        if current empty space position is different from the final order empty
+        space position
+
+        Otherwise picks up the random not arranged car and places it to the empty space
+        in this case the number of not arranged cars stays the same
+
+        Complexity is O(1)
+        '''
+
+        if self.final_car_order[self.empty_position] == 0:
+            car_to_move_position = self._get_not_arranged_car_index()
+            car_to_move = self.current_car_order[car_to_move_position]
+        else:
+            car_to_move = self.final_car_order[self.empty_position]
+            car_to_move_position = self.car_position_dict[car_to_move]
+            self.not_arranged_cars.difference_update([car_to_move]) # O(1)
+
+        self._move_car_to_empty_slot(car_to_move_position)
+        return Movement(car_to_move, car_to_move_position, self.empty_position);
+
+    def _variables_to_init_state(self):
+
+        '''Resets variables to the same state when class was initialized
+        Needed for running rearrange_all_cars function again
+        '''
+
+        self.current_car_order = self.initial_car_order
+        self.empty_position = self.current_car_order.index(0)
+        self.car_position_dict = {car:pos for pos, car in \
+                                  enumerate(self.initial_car_order) if car} # time O(n)
+        self.not_arranged_cars = {car for i, car in enumerate(self.initial_car_order)\
+                                  if self.initial_car_order[i] != self.final_car_order[i]\
+                                  and car} # time O(n)
+
+    def rearrange_all_cars(self):
+
+        '''Rearranges all cars and prints the moves
+        Note 1: In case there are more than 1000 moves to print
+        logging into the file is done
+        Note 2: Possible to run the function many times
+
+        '''
+        
+        if self.initial_car_order == self.final_car_order:
+            assert ValueError, 'Cars are already in the final order, no moves are required'
+        
+        '''Final solution is O(n)
+        '''
+
+        moves = []
+        while self.not_arranged_cars:
+            moves.append(self._rearrange_one_car()) # O(1)
+        
+        if len(moves) > 1000:
+            self.print_moves_log_file(moves)
+        else:
+            self.print_moves_console(moves)
+            
+        self._variables_to_init_state()    

```

Comment:
`self._variables_to_init_state()` (line 205) should probably be placed at the beginning of the function `rearrange_all_cars()`, otherwise I supposed all tests will fail XD